### ISSUE TITLE
Assigning node fields to NMG in read_shapefile

### ISSUE THREAD

**Issue by pfeiffea** [landlab/landlab#1138]
@mcflugen and @kbarnhart : 

I’m trying to use a shapefile to create an NMG to run NST. Issue I’m having: we need to assign topographic elevations for each node on the network. The way read_shapefile works, there isn’t an obvious way to set node elevations. Furthermore, if I just had a list of node elevations and I wanted to somehow bring them in, it gets complicated—the indexing of input links/nodes doesn’t match landlab link/node indexing. 

Jon alluded to this issue in a conversation the other day, and I’m coming to appreciate that this is a trickier issue than I’d understood. 

Is there an existing example of how to do this?

**Comment by kbarnhart** [comment]
@pfeiffea I'll work on an answer to this on the bus tonight.

**Comment by kbarnhart** [comment]
@pfeiffea I'm going to need to tap @mcflugen in here for some help. A couple of thoughts

read shape file reads a poly line type shape file, creates a network grid, and assigns any fields located on the shape file to be located _at link_. So it is unclear to me where in the shape file itself the elevation location would be stored (or do the point geometries also have z attributes?) Could you clarify on this point?

One simple approach to try, is that if you have x and y and z locations that _you know_ align with the grid nodes, then you can sort them so that the increase by x and then by y (which I think means sort first by y and then by x, but @mcflugen will correct me if I'm wrong). np.argsort and np.take_along are the two functions you'll want to check out. 

The issue you are facing with the ordering is that landlab's grid will re-order the nodes that make up the link endpoints and links based on the landlab ordering scheme. I'm pretty sure that this means that the code I wrote originally ([link to the line here](https://github.com/landlab/landlab/blob/7b15af7fb877bbf91d24c77d236863fe6a89d1b1/landlab/io/shapefile/read_shapefile.py#L171)) to assign at-link attributes to the grid is wrong and these values need to be re-ordered so that they are assigned to the correct links.

 @mcflugen could you point me in the right direction for re-ordering these values so that they can be assigned to the grid.

**Comment by mcflugen** [comment]
@kbarnhart, @pfeiffea 

There is a *sort* keyword for *NetworkGraph* that controls if the graph elements should be sorted (*landlab*-style) or not. The way things are now, *NetworkModelGrid* always sets this value to *True* (i.e. elements are always sorted).

Would one possibility be to set this value to *False* so that nothing is reordered?

If you really want things to be *landlab*-style ordering, you could create the graph with *sort=False* and then manually call the *NetworkGraph*'s *sort* method, which sorts the graph and also returns the ids of the sorted nodes and links. With those ids, you should be able to reorder the field values from the shape file.

Does that make sense? I think there might be something in there... 🤔 ?

The following won't work but maybe gives an idea of what I was thinking,
```python
    x_of_node, y_of_node = zip(*node_xy)
    grid = NetworkModelGrid((y_of_node, x_of_node), links, sort=False)
    sorted_nodes, sorted_links, sorted_patches = grid.sort()
    for field_name in fields:
        grid.at_link[field_name] = fields[field_name][sorted_links]
```

**Comment by kbarnhart** [comment]
This will definitely solve my problem. Totally made my evening.

Not 100% sure it will address the issue of elevation at nodes, but I’ll
wait to hear more details about the source of the elevations.

**Comment by pfeiffea** [comment]
Thanks, folks. 

The elevation at nodes issue is tricky. Two ideas: 
- find a way in GIS to assign an "upstream node elevation" and "downstream node elevation" to each link, then read those in after using `read_shapefile` 

-read in a point shapefile of node locations with fields of node elevations and then associate those to the grid.. (probably the easier option for the user)

Do either of you have thoughts on which is a preferable system? This seems like an issue that nearly every user of the NMG will run into, so I'm hesitant to forge ahead with a hacky Allison fix..

**Comment by mcflugen** [comment]
@pfeiffea, @kbarnhart It seems like we should have mappers that map values from links to nodes. If we do, they would be in the *mappers* module. It's not clear from their names, but maybe one of the following?
* `map_min_of_node_links_to_node`
* `map_max_of_node_links_to_node`
* `map_upwind_node_link_max_to_node`
* `map_downwind_node_link_max_to_node`
* `map_upwind_node_link_mean_to_node`
* `map_downwind_node_link_mean_to_node`
* `map_value_at_upwind_node_link_max_to_node`
* `map_value_at_downwind_node_link_max_to_node`

**Comment by pfeiffea** [comment]
Just so I'm clear: the idea would be to (e.g.) use `map_min_of_links_to_node`  to pin the shapefile link field (downstream node elevation) to node? 

The upwind/downwind seem useful in many cases, but am I right that upwind/downwind are defined via a flow director, which requires elevation?

**Comment by mcflugen** [comment]
@pfeiffea That's what I was thinking, yes. I'm not sure that any of those mappers would work for what you, though.

If we had a mapper that did the following, would it work for you: for every node, set its elevation to that of the average elevation of all links connected to it.

**Comment by pfeiffea** [comment]
@mcflugen Hmm.. i don't think so. We really just need the node elevation to be elevation of that particular point in space. We use these node elevations to calculate the average channel slopes for the links. 

That's why I think it might be easier to create an input/output tool to allow the user to bring in a point shapefile that represents nodes with a field of elevations, then map those points to the NMG grid nodes. 

The standard use case (in my imagination) would involve someone starting with a river network in GIS. They would almost certainly have a polyline shapefile of river reaches (links) and a point shapefile of reach breaks (nodes).

**Comment by mcflugen** [comment]
@pfeiffea Oh, yeah, for sure. Sorry, I thought you were stuck with having elevations defined only at links. If you have a file that defines elevations at nodes, it should be no problem to map those elevations to NMG nodes.

**Comment by pfeiffea** [comment]
@mcflugen Perfect. Yes, that's all we need!

**Comment by kbarnhart** [comment]
@pfeiffea I have a couple other things that I need to finish before early next week, but then I can spend some time on this. based on the conversation what I think makes the most sense is to:

 -[ ] Fix the current read shape file to deal with the re-ordering correctly
- [ ] add an option to read shapefile in which someone could provide a poly line shapefile and point (required to be co-located) shapefile. If the point shapefile is created, the reader will verify that that the points are co-located and then assign the point data at node. 
- [ ] We can verify that the map average value to links (or similar work) and then include it in the docstring so that it is an obvious approach as well. 

this second item will deal with the common occurrence for your planned usage. 

@pfeiffea do you have an example shapefile of links, and associated example shapefile for nodes? If so, could you send it to me.
I think the

**Comment by pfeiffea** [comment]
Hi @kbarnhart and @mcflugen... sorry for the slow response. My email was sending the GitHub messages to the junk folder. 

> add an option to read shapefile in which someone could provide a poly line shapefile and point (required to be co-located) shapefile. If the point shapefile is created, the reader will verify that that the points are co-located and then assign the point data at node. 

Yes, this sounds great. 

@kbarnhart I can create that point shapefile and send it your way.

**Comment by pfeiffea** [comment]
@kbarnhart and @mcflugen 

I'm circling back on this topic. I created a new test network including a shapefile with nodes and links. It is on my development branch under tutorials-->NST-->data.

**Comment by kbarnhart** [comment]
thanks @pfeiffea I'll try and work on this in the next week or so.

**Comment by kbarnhart** [comment]
quick update: I got about halfway through addressing this the other day. I should be able to do the other half in the next few days.

**Comment by kbarnhart** [comment]
This was addressed by #1158 closing.
